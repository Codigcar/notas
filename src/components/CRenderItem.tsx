import React from 'react';
import {Text, View, StyleSheet} from 'react-native';
import Proptypes from 'prop-types';

import {IPublication} from '../interfaces/publish.interface';

export const CRenderItem = ({title, body, userId}: IPublication) => {
  return (
    <View accessibilityRole='menuitem' style={styles.container}>
      <Text style={styles.title}>{title}</Text>
      <Text>{body}</Text>
      <Text>User: {userId}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: 10,
    borderBottomColor: '#8BA0AB',
    borderBottomWidth: 1,
  },
  title: {
    fontWeight: 'bold',
    fontSize: 20,
  },
});

CRenderItem.propTypes = {
  item: Proptypes.shape({
    id: Proptypes.number.isRequired,
    title: Proptypes.string.isRequired,
    body: Proptypes.string.isRequired,
    userId: Proptypes.number.isRequired,
  }),
};
