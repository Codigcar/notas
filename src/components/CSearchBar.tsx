import React from 'react';
import {View, Platform} from 'react-native';
import {SearchBar} from 'react-native-elements';
import {SearchBarBaseProps} from 'react-native-elements/dist/searchbar/SearchBar';
import PropTypes from 'prop-types';

 export const CSearchBar = ({
  dataList,
  search,
  setSearch,
  setFilteredDataSource,
}: any) => {
  const SafeSearchBar = SearchBar as unknown as React.FC<SearchBarBaseProps>;

  const searchFilterFunction = (text: string) => {
    if (text) {
      const newData = dataList.filter(function (item: any) {
        const itemData = item.title;
        return itemData.indexOf(text) > -1;
      });
      setFilteredDataSource(newData);
      setSearch(text);
    } else {
      setFilteredDataSource(dataList);
      setSearch(text);
    }
  };

  return (
    <View>
      <SafeSearchBar
        containerStyle={styles.containerBarraBusqueda}
        inputContainerStyle={styles.estiloBarraBusqueda}
        onChangeText={(text: string) => searchFilterFunction(text)}
        onClear={() => searchFilterFunction('')}
        placeholder="Buscar por tÃ­tulo..."
        value={search}
        platform={'android'}
        clearIcon={true}
        testID={'search'}
      />
    </View>
  );
};

export const styles = {
  containerBarraBusqueda: {
    backgroundColor: '#fff',
    borderTopColor: '#FFF',
    borderBottomColor: '#FFF',
    paddingHorizontal: 0,
    marginRight: 20,
    ...Platform.select({
      ios: {
        paddingVertical: 0,
        borderRadius: 10,
      },
    }),
  },
  estiloBarraBusqueda: {
    backgroundColor: 'white',
    borderRadius: 10,
    marginVertical: 0,
    ...Platform.select({
      ios: {},
      android: {
        marginLeft: 15,
        marginRight: 15,
        shadowOpacity: 0.39,
        shadowRadius: 13.97,
        elevation: 11,
      },
      default: {
        shadowColor: 'rgba(0,0,0, .2)',
        shadowOffset: {height: 0, width: 0},
        shadowOpacity: 1,
        shadowRadius: 1,
      },
    }),
  },
};

CSearchBar.propTypes = {
  dataList: PropTypes.array.isRequired,
  search: PropTypes.string.isRequired,
  setSearch: PropTypes.func.isRequired,
  setFilteredDataSource: PropTypes.func.isRequired,
}
