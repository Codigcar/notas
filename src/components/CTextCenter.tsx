import React from 'react';
import {View, Text, StyleSheet} from 'react-native';
import Proptypes from 'prop-types';

type Props = {
  title: string;
};

export const CTextCenter = ({title}: Props) => {
  return (
    <View style={styles.center}>
      <View>
        <Text>{title}</Text>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  center: {
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
  },
});

CTextCenter.propTypes = {
    title: Proptypes.string.isRequired
}