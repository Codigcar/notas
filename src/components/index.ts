

export * from './CRenderItem';
export * from './CRenderList';
export * from './CSearchBar';
export * from './CTextCenter';