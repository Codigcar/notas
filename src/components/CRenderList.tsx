import React from 'react';
import {FlatList, RefreshControl } from 'react-native';
import {CRenderItem} from './CRenderItem';
import PropTypes from 'prop-types';
import {CTextCenter} from './CTextCenter';

type Props = {
  data: any;
  pullRefresh: () => Promise<void>;
  isLoadingPullRefresh: boolean;
  loadMore?: (() => void) | null;
  timeEndReached?: number | null;
};

export const CRenderList = ({
  data,
  loadMore = null,
  timeEndReached = null,
  pullRefresh,
  isLoadingPullRefresh,
}: Props) => {
  return (
    <FlatList
      testID='flat-list'
      data={data}
      keyExtractor={item => item.id.toString()}
      renderItem={({item}) => <CRenderItem {...item} />}
      onEndReached={loadMore}
      onEndReachedThreshold={timeEndReached}
      refreshControl={
        <RefreshControl
          refreshing={isLoadingPullRefresh}
          onRefresh={pullRefresh}
          tintColor={'red'}
          colors={['red']}
          testID={'refresh-control'}
        />
      }
      contentContainerStyle={{flexGrow: 1}}
      ListEmptyComponent={() => {
        return <CTextCenter title="Publicaciones no encontradas" />;
      }}
    />
  );
};

CRenderList.propTypes = {
  data: PropTypes.array.isRequired,
  pullRefresh: PropTypes.func.isRequired,
  isLoadingPullRefresh: PropTypes.bool.isRequired,
  loadMore: PropTypes.func,
  timeEndReached: PropTypes.number,
};
