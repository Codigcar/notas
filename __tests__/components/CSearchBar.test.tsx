import React, {useState} from 'react';
import {CSearchBar} from '../../src/components/CSearchBar';

const {render, fireEvent, screen} = require('@testing-library/react-native');
describe('testing en <CSearchBar />', () => {
  const publicationsList = [
    {
      userId: 1,
      id: 1,
      title: 'title1',
      body: 'body1'
    },
    {
      userId: 1,
      id: 2,
      title: 'title2',
      body: 'body2'
    },
    {
      userId: 1,
      id: 3,
      title: 'title3',
      body: 'body3'
    },
  ];

  const searchText = '';
  const setSearchMock = jest.fn();
  const setFilteredDataSourceMock = jest.fn();

  beforeEach(() => jest.clearAllMocks());
  it('debe hacer match con el snapshot', () => {
    const container = render(
      <CSearchBar
        dataList={publicationsList}
        search={searchText}
        setSearch={setSearchMock}
        setFilteredDataSource={setFilteredDataSourceMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('debe ejecutarse las funcion setSearch con nuevo texto ingresado', () => {
    const changeTextTo = 'title2';

    const component = render(
      <CSearchBar
        dataList={publicationsList}
        search={searchText}
        setSearch={setSearchMock}
        setFilteredDataSource={setFilteredDataSourceMock}
      />,
    );
    const textInput = component.getByTestId('search');

    fireEvent.changeText(textInput, changeTextTo);

    expect(setSearchMock).toHaveBeenCalled();
    expect(setSearchMock).toHaveBeenCalledWith(changeTextTo);
  });

  it('debe ejecutarse las funcion setFilteredDataSource con el nuevo array de publicaciones segÃºn el texto ingresado', () => {
    const changeTextTo = 'title2';

    const component = render(
      <CSearchBar
        dataList={publicationsList}
        search={searchText}
        setSearch={setSearchMock}
        setFilteredDataSource={setFilteredDataSourceMock}
      />,
    );
    const textInput = component.getByTestId('search');

    fireEvent.changeText(textInput, changeTextTo);

    const newData = publicationsList.filter(publication => publication.title == changeTextTo)

    expect(setFilteredDataSourceMock).toHaveBeenCalled();
    expect(setFilteredDataSourceMock).toHaveBeenCalledWith(newData);
  });
});
