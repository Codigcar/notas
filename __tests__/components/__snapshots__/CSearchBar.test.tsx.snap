// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing en <CSearchBar /> debe hacer match con el snapshot 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderBottomColor": "#FFF",
        "borderRadius": 10,
        "borderTopColor": "#FFF",
        "marginRight": 20,
        "paddingBottom": 8,
        "paddingHorizontal": 0,
        "paddingTop": 8,
        "paddingVertical": 0,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 10,
            "flexDirection": "row",
            "marginVertical": 0,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "rgba(0, 0, 0, 0.54)",
                        "fontSize": 25,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="iconIcon"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          editable={true}
          onBlur={[Function]}
          onCancel={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Buscar por título..."
          placeholderTextColor="#86939e"
          platform="android"
          style={
            Object {
              "alignSelf": "center",
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 24,
              "marginRight": 8,
              "minHeight": 40,
            }
          }
          testID="search"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;
